examples/atari/README.md

```diff
@@ -82,6 +82,20 @@ One epoch here is equal to 100,000 env step, 100 epochs stand for 10M.
 | SeaquestNoFrameskip-v4      | 10775       | ![](results/fqf/Seaquest_rew.png)     | `python3 atari_fqf.py --task "SeaquestNoFrameskip-v4"`  |
 | SpaceInvadersNoFrameskip-v4 | 2482        | ![](results/fqf/SpaceInvaders_rew.png) | `python3 atari_fqf.py --task "SpaceInvadersNoFrameskip-v4"`  |
 
+# Rainbow (single run)
+
+One epoch here is equal to 100,000 env step, 100 epochs stand for 10M.
+
+| task                        | best reward | reward curve                          | parameters                                                   |
+| --------------------------- | ----------- | ------------------------------------- | ------------------------------------------------------------ |
+| PongNoFrameskip-v4          | 20.0        | ![](results/rainbow/Pong_rew.png)         | `python3 atari_rainbow.py --task "PongNoFrameskip-v4" --batch-size 64` |
+| BreakoutNoFrameskip-v4      | 315.8       | ![](results/rainbow/Breakout_rew.png)     | `python3 atari_rainbow.py --task "BreakoutNoFrameskip-v4" --n-step 1` |
+| EnduroNoFrameskip-v4        | 461         | ![](results/rainbow/Enduro_rew.png)       | `python3 atari_rainbow.py --task "EnduroNoFrameskip-v4"`  |
+| QbertNoFrameskip-v4         | 4137.5      | ![](results/rainbow/Qbert_rew.png)        | `python3 atari_rainbow.py --task "QbertNoFrameskip-v4"`  |
+| MsPacmanNoFrameskip-v4      | 2419        | ![](results/rainbow/MsPacman_rew.png)     | `python3 atari_rainbow.py --task "MsPacmanNoFrameskip-v4"`  |
+| SeaquestNoFrameskip-v4      | 590         | ![](results/rainbow/Seaquest_rew.png)     | `python3 atari_rainbow.py --task "SeaquestNoFrameskip-v4"`  |
+| SpaceInvadersNoFrameskip-v4 | 645.5       | ![](results/rainbow/SpaceInvaders_rew.png) | `python3 atari_rainbow.py --task "SpaceInvadersNoFrameskip-v4"`  |
+
```

Comment:
I've been trying to figure this out for the past week. Fixed a few bugs but the results are still pretty bad. A few observations:
- Unlike other methods, where the test reward climbs up, there is a lot of thrashing during the training, i.e., a high test reward followed by a very low test reward then high again;
- The training loss is so small that I'm worried that we might be encountering numerical issues. This is due to the fact that our C51 implementation computes `softmax` instead of `log_softmax`. See reference https://github.com/Kaixhin/Rainbow/blob/9ff5567ad1234ae0ed30d8471e8f13ae07119395/agent.py#L66 . However, it seems the C51 results are still quite reasonable so I'm not sure.

I attached a screenshot of the symptoms. What do you think?

<img width="1076" alt="Screen Shot 2021-07-08 at 9 02 19 AM" src="https://user-images.githubusercontent.com/10931178/124954766-4a218d00-dfcb-11eb-8de7-67a1786f3dec.png">

Reply:
have you tried both prio-buffer and normal buffer?

Reply:
You might be onto something. I tried to disable one aspect of Rainbow and got the following:

- disable dueling: 3797.5 -> 3070.0
- disable noisy net: 3797.5 -> 3850.0
- disable prio-buffer: 3797.5 -> 7830.0

So it seems that prio-buffer is making things worse. I previously did a sanity check run where I disabled all three and got back the C51 results. I'll do it again since I've changed some code since then.

Have you checked the correctness of prio-buffer before? Do you think there might be a bug or just hyper parameter misalignment? Currently I'm using `--alpha 0.5 --beta 1`. Recall we had a discussion above that the paper and the reference code mentioned annealing `beta` from 0.4 to 1 but my previous experiments didn't show any visible difference.

Reply:
I'll check it tonight. Btw, have you tried to use only 1 env (and using PrioReplayBuffer instead of vectorized version)?

Reply:
No, I haven't. I'll give it a try.

Reply:
Btw if you have tried to align with sb2 prio-buffer, here's the issue: https://github.com/hill-a/stable-baselines/issues/941

Reply:
I think it is related to 3 aspects:

1. policy's learn function `batch.weight` calculation, to send to `buffer.update_weight`;
2. single prio-buffer's correctness and performance;
3. vectorized prio-buffer's correctness and performance.

Reply:
I ran several experiments over the weekend and here are the results on Qbert so far:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| current rainbow | Y | Y | Y | 3797.5 |
| no prio                 | N | Y | Y  | 7830   |
| no noisy               | Y | N | Y | 3850 |
| no dueling          | Y | Y | N | 3070 |
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| single prio           | Y | Y | Y | 4240 |
| single prio only   | Y | N | N | 4962.5 |

My conclusions:
- Priority buffer is the main culprit; from "sanity check" (=C51) to "prio only" (=C51 w/ prio): 16530 -> 3982.5;
- Single vs vectorized might have some difference but should not be the main issue.

I'll try removing the eps hack and see what happens.

Reply:
I got some results back. See below:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| w/o eps hack (beta=1) | Y | N | N | 770 |
| w/o eps hack + beta=0.7 | Y | N | N | 3950 |
| w/o eps hack + beta=0.4 | Y | N | N | 4957.5 |
| single prio only   | Y | N | N | 4962.5 |
| w/o eps hack (beta=1) | Y | N | N | 692.5 |
| current rainbow | Y | Y | Y | 3797.5 |
| w/o eps hack (beta=1) | Y | Y | Y | 1000 |
| w/o eps hack + beta=0.4 | Y | Y | Y | 5090 |

I don't know why "removing eps hack + beta=1" got the worst result but "removing eps hack + beta=0.4" is currently the best. Single prio buffer's results seemed to follow vectorized prio buffer closely.

I'm running more experiments with noisy_std=0.5 (currently noisy_std=0.1 following Kaixhin's setup).

Reply:
I have another idea: instead of getting rid of eps hack, which protects against div-by-zero exception, I can scale the weights by a constant, such as `1e9` so that adding eps (~`1e-7`) wouldn't affect the weights significantly.

Reply:
But have you carefully examined the prio-buffer and align with other (rainbow) successful implementations? (Not sure what's happening but I've tried my best to check the buffer is correct)

And one more thing: have you use single env with update_freq= 4 env_step?

Reply:
> But have you carefully examined the prio-buffer and align with other (rainbow) successful implementations? (Not sure what's happening but I've tried my best to check the buffer is correct)
> 
No, I haven't. I believe that you most likely know more about it than I do. If you can't find any major error, neither can I.

> And one more thing: have you use single env with update_freq= 4 env_step?
Do you mean `atari_rainbow.py --target-update-freq=4`? (current default 500.)

Reply:
No i mean currently 10 env step once and perform update, that's 10:1 env_step:grad_step, and in traditional dqn setting it is 4:1

Reply:
> If you can't find any major error, neither can I.

Don't say that, the more people check, the more reliable the code is.

Reply:
> No i mean currently 10 env step once and perform update, that's 10:1 env_step:grad_step, and in traditional dqn setting it is 4:1

I see. So it's `atari_rainbow.py --training-num 4`, right? Sure I can try it.

Reply:
Weight scaling seems to be working very well:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| w/ weight scaling + beta=0.4 | Y | N | N | 16582.5 |
| weight scaling + beta=0.4 | Y | Y | Y |  11625  |

Reply:
> I see. So it's atari_rainbow.py --training-num 4, right? Sure I can try it.

Yep